www/docs/reactjs/mocking.md

```diff
@@ -0,0 +1,215 @@
+---
+id: mocking
+title: Mocking your server
+sidebar_label: Mocking your server
+slug: /reactjs/mocking
+---
+
+Mocking tRPC gives you the ability to isolate your frontend components into testing or presentation environments without having to alter how they behave.
+
+By intercepting our tRPC calls at the [Service Worker level](https://developer.mozilla.org/en-US/docs/Web/API/Service_Worker_API) and returning responses as the component expects, mocking can help you to:
+
+- Write performant, easy-to-execute tests that don't rely on changing the underlying component in any way
+Present components in frontend workshops like Storybook exactly as they would appear and behave in your main app
+
+The examples below use [Vitest](https://vitest.dev/), [React Testing Library](https://github.com/testing-library/react-testing-library), and [Mock Service Worker](https://www.npmjs.com/package/msw-trpc) (MSW), but the concepts apply to any testing framework.
+
+## Creating a MockedTRPCProvider
+
+For tRPC's hooks to work, they require [context](https://trpc.io/docs/context) that describes the URL of the backend and the structure of the API.
+
+- In React, you achieve this by wrapping your component tree with the `trpc.Provider` and `QueryClientProvider` components.
+- In Next.js, this step is abstracted in `_app.ts` by using `trpc.withTRPC(MyApp)`.
+
+We still need this context when using our components in an isolated environment, because they still need to be capable of making requests that we will later intercept and pass through the mocked version of our API. For this, we will need to create and use a mocked provider.
+
+### 1. Initialize a new React hook
+
+Create a new file for your `MockedTRPCProvider` and import your `AppRouter` type from where your standard tRPC router is defined. This will ensure that your mocks have the same type signature as your API.
+
+```ts title='mockedTRPCProvider.ts'
+import type { AppRouter } from "../path/to/server/trpc";
+```
+
+Next, create a set of React hooks from your `AppRouter` type signature with `createTRPCReact`.
+
+```ts title='mockedTRPCProvider.ts'
+const mockedTRPC = createTRPCReact<AppRouter>({
+  unstable_overrides: {
+    useMutation: {
+      async onSuccess(opts) {
+        await opts.originalFn();
+        await opts.queryClient.invalidateQueries();
+      },
+    },
+  },
+});
+```
+
+### 2. Create your mocked tRPC provider
+
+Now we just need to create a wrapper that exposes the client. This process is similar to [the React setup](https://trpc.io/docs/react#3-add-trpc-providers) however we don't need to utilize the `useState` hook because our tests are not using SSR.
+
+```ts title='mockedTRPCProvider.ts'
+// @ts-expect-error tRPC utilizes global fetch
+global.fetch = fetch;
+
+const mockedTRPCClient = mockedTRPC.createClient({
+  links: [httpBatchLink({ url: "http://localhost:3000/api/trpc" })],
```

Comment:
`createClient` also needs to be passed a transformer
(note: this is also missing in trpc/react-query docs pages. maybe a recent type change?)

Reply:
Since transformer is optional, isn't it best to keep the docs boiled down to their simplest example? I think that's been the way so far in the docs and reduces visual noise for new learners

Reply:
as far as i can tell its not possible to not pass a transformer here 
<img width="689" alt="image" src="https://user-images.githubusercontent.com/8353666/226212852-620ff55d-0b53-4d52-8c42-bfc136715662.png">